File: content_tool/agents/topic_hot.py
# ruff: noqa: RUF001
"""Topic hot-topic agent.

One Gemini call per candidate. Inspects the HK SERP for the topic and decides
whether it qualifies as a "hot topic." No retry/backoff here — that is the
topic-expansion subgraph's concern (Task 3).
"""

from sqlalchemy.ext.asyncio import AsyncSession

from content_tool import prompts_store
from content_tool.gemini.client import GeminiClient
from content_tool.models.topic_batch import TopicHotInput, TopicHotOutput
from content_tool.policy.personas import load_persona

# HK-ZH default market (mirrors VoiceLocale.market). Keeps the assembled prompt
# byte-identical for bowtie-editor when no voice locale override is present.
_DEFAULT_MARKET = "Google 香港繁中"


async def build_system_prompt(voice_slug: str = "bowtie-editor") -> str:
    return await prompts_store.get_assembled_standalone("topic_hot", voice_slug=voice_slug)


def build_user_prompt(input_: TopicHotInput, market: str = _DEFAULT_MARKET) -> str:
    keywords = ", ".join(input_.keywords) if input_.keywords else "（無）"
    return (
        f"請分析以下單一 topic 在 {market} SERP 是否屬於熱門話題。"
        "只輸出符合 schema 的 JSON。\n\n"
        f"topic:\n{input_.topic}\n\n"
        f"focus_keywords:\n{keywords}\n"
    )


async def run_topic_hot(
    *,
    gemini: GeminiClient,
    input: TopicHotInput,
    voice_slug: str = "bowtie-editor",
    session: AsyncSession | None = None,
) -> TopicHotOutput:
    """Single Gemini call. Returns the hot-topic verdict for one candidate.

    When ``session`` is supplied, the voice's :class:`VoiceLocale.market` is
    threaded into the user prompt (spec §4.4 item 7) so a non-HK voice asks about
    its own market. Without a session the HK-ZH default keeps the prompt
    byte-identical.
    """
    system_prompt = await build_system_prompt(voice_slug)
    market = _DEFAULT_MARKET
    if session is not None:
        pack = await load_persona(voice_slug, session=session)
        market = pack.locale.market
    user_prompt = build_user_prompt(input, market)
    result = await gemini.generate(
        agent="topic_hot",
        system_prompt=system_prompt,
        user_prompt=user_prompt,
        response_schema=TopicHotOutput.model_json_schema(),
        tools=["googleSearch", "urlContext"],
    )
    return TopicHotOutput.model_validate(result.parsed)

File: deploy/cloudflare-workers/src/agents/topic_hot.ts
/**
 * Topic hot-topic agent — TypeScript port of `content_tool/agents/topic_hot.py`.
 *
 * One Gemini call per candidate: inspects the HK SERP for the topic and decides
 * whether it qualifies as a "hot topic." No retry/backoff and no DB writes —
 * those are the TopicExpansionWorkflow's concern (assembled by the lead). This
 * function returns the parsed verdict plus token usage.
 *
 * Full-width punctuation (：、（無）) is intentional — the prompts are CJK style,
 * matching the Python `build_user_prompt` byte-for-byte.
 */

import type { Sql } from "postgres";
import { getAssembled } from "../prompts/store";
import { loadPersona } from "./persona";
import type { GeminiClient, ThoughtCallback } from "../gemini/types";
import { TOPIC_HOT_SCHEMA, type TopicHotOutput } from "./topic_schemas";

// ---------------------------------------------------------------------------
// Input — mirrors Python TopicHotInput.
// ---------------------------------------------------------------------------

export interface TopicHotInput {
  topic: string;
  keywords: string[];
  /** The batch/candidate voice (persona slug); resolves the prompt under it. */
  voiceSlug: string;
  onThought?: ThoughtCallback;
}

// ---------------------------------------------------------------------------
// Token usage result (mirrors OutlineTokens)
// ---------------------------------------------------------------------------

export interface TopicHotTokens {
  tokensIn: number;
  tokensOut: number;
  thinkingTokens: number;
  latencyMs: number;
}

// ---------------------------------------------------------------------------
// System prompt assembly — mirrors Python `build_system_prompt`.
// ---------------------------------------------------------------------------

async function buildSystemPrompt(sql: Sql, voiceSlug: string): Promise<string> {
  return getAssembled(sql, "topic_hot", voiceSlug);
}

// ---------------------------------------------------------------------------
// User prompt — mirrors Python `build_user_prompt` exactly.
// ---------------------------------------------------------------------------

export function buildUserPrompt(opts: {
  topic: string;
  keywords: string[];
  /** The voice's market (VoiceLocale.market); defaults to HK-ZH "Google 香港繁中". */
  market?: string;
}): string {
  const keywords = opts.keywords.length > 0 ? opts.keywords.join(", ") : "（無）";
  const market = opts.market ?? "Google 香港繁中";
  return (
    `請分析以下單一 topic 在 ${market} SERP 是否屬於熱門話題。` +
    "只輸出符合 schema 的 JSON。\n\n" +
    `topic:\n${opts.topic}\n\n` +
    `focus_keywords:\n${keywords}\n`
  );
}

// ---------------------------------------------------------------------------
// Main export
// ---------------------------------------------------------------------------

/**
 * Run the topic-hot agent for one candidate. Assembles the system + user prompt,
 * calls Gemini with structured output and the googleSearch + urlContext
 * grounding tools, and returns the parsed verdict plus token usage. No DB writes.
 */
export async function runTopicHot(
  sql: Sql,
  gemini: GeminiClient,
  input: TopicHotInput,
): Promise<{ output: TopicHotOutput; tokens: TopicHotTokens }> {
  const systemPrompt = await buildSystemPrompt(sql, input.voiceSlug);
  // Resolve the voice's market locally (spec §4.4 item 7) so a non-HK voice asks
  // about ITS market. HK-ZH default keeps "Google 香港繁中" → byte-identical.
  const { locale } = await loadPersona(sql, input.voiceSlug);
  const userPrompt = buildUserPrompt({
    topic: input.topic,
    keywords: input.keywords,
    market: locale.market,
  });

  const result = await gemini.generate({
    agent: "topic_hot",
    systemPrompt,
    userPrompt,
    responseSchema: TOPIC_HOT_SCHEMA as Record<string, unknown>,
    tools: ["googleSearch", "urlContext"],
    onThought: input.onThought,
  });

  // Gemini returns a plain object matching TopicHotOutput's shape when
  // responseSchema is provided. Route through `unknown` since
  // `GeminiResult.parsed` is `Record<string, unknown>`.
  const output = result.parsed as unknown as TopicHotOutput;

  return {
    output,
    tokens: {
      tokensIn: result.tokensIn,
      tokensOut: result.tokensOut,
      thinkingTokens: result.thinkingTokens,
      latencyMs: result.latencyMs,
    },
  };
}
